### PR TITLE
Stream refac and vertex streaming

### PR DESCRIPTION
**Pull request by baskaryan**
will add async if we like this cc @eyurtsev @nfcampos

### CODE REVIEW

File: libs/langchain/langchain/chat_models/konko.py [hunk 1]
```diff
@@ -224,16 +224,9 @@ def _generate(
...
-            for chunk in self._stream(
-                messages=messages, stop=stop, run_manager=run_manager, **kwargs
-            ):
-                if generation is None:
-                    generation = chunk
-                else:
-                    generation += chunk
-            assert generation is not None
-            return ChatResult(generations=[generation])
+            return self._generate_from_stream(
```
Comment (eyurtsev) [inline libs/langchain/langchain/chat_models/konko.py:227]: are we able to remove `stream` from the method signature?

File: libs/langchain/langchain/chat_models/vertexai.py [hunk 2]
```diff
@@ -91,10 +100,23 @@ def _parse_examples(examples: List[BaseMessage]) -> List["InputOutputTextPair"]:
     return example_pairs
 
 
+def _get_question(messages: List[BaseMessage]) -> HumanMessage:
+    """Get the human message at the end of a list of input messages to a chat model."""
+    if not messages:
+        raise ValueError("You should provide at least one message to start the chat!")
+    question = messages[-1]
+    if not isinstance(question, HumanMessage):
```
Comment (eyurtsev) [inline libs/langchain/langchain/chat_models/vertexai.py:108]: Are there no situations when we want to continue generation given the last message was actually AI? Or maybe it was just the system message?

File: libs/langchain/langchain/chat_models/vertexai.py [hunk 3]
```diff
@@ -118,6 +140,7 @@ def _generate(
         messages: List[BaseMessage],
         stop: Optional[List[str]] = None,
         run_manager: Optional[CallbackManagerForLLMRun] = None,
+        stream: Optional[bool] = False,
```
Comment (eyurtsev) [inline libs/langchain/langchain/chat_models/vertexai.py:143]: if we're able to remove 
```suggestion
```

File: libs/langchain/langchain/chat_models/vertexai.py [hunk 4]
```diff
@@ -134,27 +157,52 @@ def _generate(
...
+        for response in responses:
+            text = self._enforce_stop_words(response.text, stop)
+            if run_manager:
+                run_manager.on_llm_new_token(text)
+            yield ChatGenerationChunk(message=AIMessageChunk(content=text))
+
+    def _start_chat(
+        self, history: _ChatHistory, params: dict
+    ) -> Union[ChatSession, CodeChatSession]:
+        if not self.is_codey_model:
```
Comment (eyurtsev) [inline libs/langchain/langchain/chat_models/vertexai.py:203]: the underlying function checks by string matching 'code' -- probably low chance of false positive but feels a bit dangerous

File: libs/langchain/langchain/chat_models/base.py [hunk 5]
```diff
@@ -523,6 +523,24 @@ async def _agenerate(
...
+    ) -> ChatResult:
+        generation: Optional[ChatGenerationChunk] = None
+        for chunk in self._stream(
+            messages=messages, stop=stop, run_manager=run_manager, **kwargs
+        ):
+            if generation is None:
+                generation = chunk
+            else:
+                generation += chunk
+        assert generation is not None
```
Comment (eyurtsev) [inline libs/langchain/langchain/chat_models/base.py:541]: I generally prefer avoiding the assert statement and instead raise AssertionErrors, to stick to a style of raising exceptions

**Review by eyurtsev — APPROVED** [review]

File: libs/langchain/langchain/chat_models/vertexai.py (hunk 2 above)
Comment (baskaryan) [reply libs/langchain/langchain/chat_models/vertexai.py:108]: cc @lkuligin

File: libs/langchain/langchain/chat_models/vertexai.py (hunk 3 above)
Comment (baskaryan) [reply libs/langchain/langchain/chat_models/vertexai.py:143]: fwiw we do have it for other models, tho i agree i don't love it. also seeing default val here is wrong ,should be None

File: libs/langchain/langchain/chat_models/vertexai.py (hunk 4 above)
Comment (baskaryan) [reply libs/langchain/langchain/chat_models/vertexai.py:204]: agreed (tho matter for a sep pr probably)

File: libs/langchain/langchain/chat_models/vertexai.py (hunk 2 above)
Comment (lkuligin) [reply libs/langchain/langchain/chat_models/vertexai.py:108]: I doubt it's the case for a chat model. And Vertex chat model exposes a `send_message` method, so you can't actually send an AI message there.
https://cloud.google.com/python/docs/reference/aiplatform/latest/vertexai.language_models.ChatSession#vertexai_language_models_ChatSession_send_message

File: libs/langchain/langchain/chat_models/vertexai.py (hunk 3 above)
Comment (nfcampos) [reply libs/langchain/langchain/chat_models/vertexai.py:143]: what we shouldn't have is the instance one

**Pull request merged**